include/seqan3/alignment/scoring/detail/simd_matrix_scoring_scheme.hpp

```diff
@@ -0,0 +1,145 @@
+{
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr simd_matrix_scoring_scheme() = default; //!< Defaulted.
+    constexpr simd_matrix_scoring_scheme(simd_matrix_scoring_scheme const &) = default; //!< Defaulted.
+    constexpr simd_matrix_scoring_scheme(simd_matrix_scoring_scheme &&) = default; //!< Defaulted.
+    constexpr simd_matrix_scoring_scheme & operator=(simd_matrix_scoring_scheme const &) = default; //!< Defaulted.
+    constexpr simd_matrix_scoring_scheme & operator=(simd_matrix_scoring_scheme &&) = default; //!< Defaulted.
+    ~simd_matrix_scoring_scheme() = default; //!< Defaulted.
+
+    //!\copydoc seqan3::detail::simd_matrix_scoring_scheme::initialise_from_scalar_scoring_scheme
+    constexpr explicit simd_matrix_scoring_scheme(scoring_scheme_t const & scoring_scheme)
+    {
+        initialise_from_scalar_scoring_scheme(scoring_scheme);
+    }
+
+    //!\copydoc seqan3::detail::simd_matrix_scoring_scheme::initialise_from_scalar_scoring_scheme
+    constexpr simd_matrix_scoring_scheme & operator=(scoring_scheme_t const & scoring_scheme)
+    {
+        initialise_from_scalar_scoring_scheme(scoring_scheme);
+        return *this;
+    }
+    //!\}
+
+    /*!\name Score computation
+     *\{
+     */
+    /*!\brief Given two simd vectors, compute an element-wise score and return in another simd vector.
+     * \param[in] lhs The left operand to compare.
+     * \param[in] rhs The right operand to compare.
+     * \returns A simd vector of same type as input, with scores computed between both inputs.
+     *
+     * ### Exception
+     *
+     * No-throw guarantee.
+     *
+     * ### Complexity
+     *
+     * Linear in the length of one input vector. (`lhs` and `rhs` are equally sized)
+     *
+     * ### Thread safety
+     *
+     * Thread-safe.
+     */
+    constexpr simd_score_t score(simd_score_t const & lhs, simd_score_t const & rhs) const noexcept
+    {
+        simd_score_t result;
+
+        for (int i = 0; i < simd_traits<simd_score_t>::length; i++)
+        {
+            if (is_padded(lhs[i]) || is_padded(rhs[i]))
+            {
+                result[i] = std::same_as<alignment_t, detail::global_alignment_type> ? 1 : -1;
+            }
+            else
+            {
+                result[i] = internal_scoring_scheme.score(assign_rank_to(lhs[i], alphabet_t{}),
+                                                          assign_rank_to(rhs[i], alphabet_t{}));
+            }
+        }
+
+        return result;
+    }
+    //!\}
+
+private:
+    //!\brief Internally stores the given scalar scoring scheme matrix.
+    scoring_scheme_t internal_scoring_scheme;
+
+    /*!\brief Store the given scoring scheme matrix into a private member variable.
+     * \param[in] scoring_scheme The scoring scheme to initialise the vectorised match and mismatch score with.
+     */
+    constexpr void initialise_from_scalar_scoring_scheme(scoring_scheme_t const & scoring_scheme)
```

Comment:
why do we have this private member function if this is only doing an assignment? I don't see the benefit of abstraction here.

Reply:
I copied it over from @rrahn 's pull request, and kept it in because I thought we might want the format to be consistent... Does it not matter?

Reply:
@rrahn Should I mirror the format that you had in your pull request, with `initialize_from_scalar_scoring_scheme`? Or should I just put the `internal_scoring_scheme = scoring_scheme` assignment in the constructor/assignment operator?

Reply:
No, if the scoring scheme is the only member that is initialised you can actually use the direct member initialiser in the constructor. For the assignment operator you assign it of course in the body.

Reply:
But hold on. You actually need to test if the score stored in the scoring scheme does not exceed the largest value that is representable with the scalar type of the simd vector. So whenever the score type of the underlying scoring scheme is smaller or equal in the byte size this is not a problem. But if you have a byte packing in the simd vector and you use int16_t as a score type for the scoring scheme an overflow is possible. Thus, this should be tested during the construction. This would then again validate to move this function out.

Reply:
How would that look? Maybe something like this...?
```cpp
        static_assert(scoring_scheme_t::score_type <= simd_traits<simd_score_t>::scalar_type);

```cpp

But I don't know how to do this correctly... Do I get the max possible value of `score_type` and `scalar_type` and compare them?

Reply:
Ah wait I see you did it in yours.